### ISSUE TITLE
Rule proposal: declare variables before/after usage

### ISSUE THREAD

**Issue by ehmicky** [eslint/eslint#13144]
**Please describe what the rule should do:**

Developers usually abide to one of the following two styles when it comes to variable declaration/usage.

Declare variables before usage:

```js
const CONSTANT = true
const identity = function(value) {
  return value
}

const exampleFunc = function() {
  return identity(CONSTANT)
}
```

Declare variables after usage:

```js
const exampleFunc = function() {
  return identity(CONSTANT)
}

const identity = function(value) {
  return value
}
const CONSTANT = true
```

Both are valid but they require reading source files differently: either from top to bottom or from bottom to top. 

Enforcing a specific style could help developers maintain consistency in how their code is read and written.

Note that this is only relevant when the variable is declared is an upper scope. If the variable is declared in the same scope, it has to be declared first:

```js
const CONSTANT = true
const OTHER_CONSTANT = CONSTANT
```

Note that this is different from the [`no-use-before-define` rule](https://eslint.org/docs/rules/no-use-before-define) which is only about hoisting.

Ideally this rule would be fixable too.

Note: I think in some cases, this rule might not be enforceable, for example when using circular references: 

```js
const one = function() {
  return two()
}

const two = function() {
  return one()
}
```

The rule would ignore those cases.

**What category of rule is this? (place an "X" next to just one item)**

[ ] Warns about a potential error (problem)
[ ] Suggests an alternate way of doing something (suggestion)
[x] Enforces code style (layout)
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

See the 2 code examples above.

**Why should this rule be included in ESLint (instead of a plugin)?**

It deals with a core styling issue with how JavaScript developers write their code.

**Comment by mdjermanovic** [comment]
Hi @ehmicky, thanks for the proposal!

> Declare variables before usage:
> 
> ```js
> const CONSTANT = true
> const identity = function(value) {
>   return value
> }
> 
> const exampleFunc = function() {
>   return identify(CONSTANT)
> }
> ```

Can you please clarify how is this different from `no-use-before-define` with `{ "variables": true }` (default)?

**Comment by ehmicky** [comment]
`no-use-before-`define` with `{ "variables": true }` does not enforce the above style. For example, it does not warn against the reverse order:

```js
const exampleFunc = function() {
  return identity(CONSTANT)
}

const identity = function(value) {
  return value
}
const CONSTANT = true
```

Also `no-use-before-define` cannot be used for developers that wish the enforce the style from the example right above (top to bottom referencing instead of bottom to top).

It seems that `no-use-before-define` is mostly preventing against hoisting and invalid syntax. My proposal would be more about whether functions/constants are written top to bottom or bottom to top in a file.

**Comment by mdjermanovic** [comment]
> `no-use-before-define` with `{ "variables": true }` does not enforce the above style. For example, it does not warn against the reverse order:
> 
> ```js
> const exampleFunc = function() {
>   return identity(CONSTANT)
> }
> 
> const identity = function(value) {
>   return value
> }
> const CONSTANT = true
> ```

`no-use-before-define` reports errors on `identity` and `CONSTANT` references, as in [this demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXVzZS1iZWZvcmUtZGVmaW5lOiBbXCJlcnJvclwiLCB7IFwidmFyaWFibGVzXCI6IHRydWUgfV0gKi9cblxuY29uc3QgZXhhbXBsZUZ1bmMgPSBmdW5jdGlvbigpIHtcbiAgcmV0dXJuIGlkZW50aXR5KENPTlNUQU5UKVxufVxuXG5jb25zdCBpZGVudGl0eSA9IGZ1bmN0aW9uKHZhbHVlKSB7XG4gIHJldHVybiB2YWx1ZVxufVxuY29uc3QgQ09OU1RBTlQgPSB0cnVlIiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjo2LCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7fSwiZW52Ijp7fX19).

Can you please provide an example that is valid for `no-use-before-define` but would be invalid for the new rule (the first style)?

**Comment by ehmicky** [comment]
Oh, sorry, yes you're right, this does enforce the first style.

So the rule would instead be enforcing the second style, i.e. top-to-bottom.

Valid:

```js
const exampleFunc = function() {
  return identity(CONSTANT)
}

const identity = function(value) {
  return value
}
const CONSTANT = true
```

Invalid:

```js
const CONSTANT = true
const identity = function(value) {
  return value
}

const exampleFunc = function() {
  return identity(CONSTANT)
}
```

**Comment by mdjermanovic** [comment]
Thanks for clarifying this!

Regarding the second style (top-to-bottom), I'm not sure how common it is. Enforcing this could lead to TDZ errors in some cases.

For example, the rule warns in this case:

```js
const CONSTANT = true
const identity = function(value) {
  return value
}

const exampleFunc = function() {
  return identity(CONSTANT)
}
foo(exampleFunc);
```

if the user prefers to manually fix code issues, this could become:

```js
const exampleFunc = function() {
  return identity(CONSTANT)
}
foo(exampleFunc); // possible error

const CONSTANT = true
const identity = function(value) {
  return value
}
```

Is this style recommended by any popular style guides?

**Comment by ehmicky** [comment]
I don't think I have read any style guide that enforces either way.

I do agree that the top-to-bottom approach increases the chances of missing references as you point out. 

On the other hand, with bigger functions, the bottom-to-top approach leads to a rather unnatural way of reading a source file: functions are read from the bottom of the file to the top, but each function body is read from its top to its bottom. This leads to switching between reading top-to-bottom and bottom-to-top. This also does not follow the way we usually read text in non-programming contexts.

Now you do make a good point about the potential for missing references with this approach if not careful, so I understand if that does not fit the requirements for a new rule. :+1:

**Comment by kaicataldo** [comment]
I don't think this meets the high bar we have to set of "ESLint is incomplete without this added rule". Given the challenges described by @mdjermanovic and it not being a style recommended by a popular style guide, this seems like a better fit for a custom rule if this is something you'd like to enforce.

**Comment by ehmicky** [comment]
Alright, thanks for following up on this!
